### PR TITLE
[src] CudaDecoder endpointing

### PR DESCRIPTION
**Pull request by hugovbraun**
Built on top of #4101. Probably better to look at the diff once #4101 has been merged.

Implements endpointing directly into the cuda decoder. 

Uses the rules as defined in online2/online-endpoint.h. From a user point of view, setting the high level parameters on endpointing and passing a vector in the DecodeBatch should be enough:

```
  void DecodeBatch(const std::vector<CorrelationID> &corr_ids,
                   const std::vector<SubVector<BaseFloat>> &wave_samples,
                   const std::vector<bool> &is_first_chunk,
                   const std::vector<bool> &is_last_chunk,
                   std::vector<std::string *> *partial_hypotheses = NULL,
                   std::vector<bool> *end_points = NULL)
```
Parameters (from online2/online-endpoint.h):

```
--endpoint.rule1.max-relative-cost : This endpointing rule requires relative-cost of final-states to be <= this value (describes how good the probability of final-states is). (float, default = inf)
--endpoint.rule1.min-trailing-silence : This endpointing rule requires duration of trailing silence(in seconds) to be >= this value. (float, default = 5)
--endpoint.rule1.min-utterance-length : This endpointing rule requires utterance-length (in seconds) to be >= this value. (float, default = 0)
--endpoint.rule1.must-contain-nonsilence : If true, for this endpointing rule to apply there mustbe nonsilence in the best-path traceback. (bool, default = false)
```

Easiest way to test is to pass `--print-endpoints=true` to the binary `src/cudadecoderbin/batched-wav-nnet3-cuda-online`

Internally, it provides all the necessary metrics (relative cost, number of silence phones on current best path, total length). Those rules can be modified through the command line parameters.

### CODE REVIEW

**Comment by kkm000** [comment]
Thanks! So, you'd say that both #4101 and this one are ready to merge?

File: src/cudadecoder/batched-threaded-nnet3-cuda-online-pipeline.cc [hunk 1]
```diff
@@ -250,6 +256,8 @@ void BatchedThreadedNnet3CudaOnlinePipeline::DecodeBatch(
         "You need to set --word-symbol-table to use partial hypotheses");
     cuda_decoder_->AllowPartialHypotheses();
   }
+  if (end_points) cuda_decoder_->AllowEndPointing();
```
Comment (kkm000) [inline src/cudadecoder/batched-threaded-nnet3-cuda-online-pipeline.cc:259]: Naming of functions varies between `Endpoint` and `EndPoint`. We generally use "endpoint" as a single word, so the former would be better.
```suggestion
  if (end_points) cuda_decoder_->AllowEndpointing();
```

File: src/cudadecoder/cuda-decoder.h [hunk 2]
```diff
@@ -236,14 +241,30 @@ class CudaDecoder {
...
 
-  void AllowPartialHypotheses() { generate_partial_hypotheses_ = true; }
+  void AllowPartialHypotheses() {
+    partial_traceback_ = generate_partial_hypotheses_ = true;
+  }
+
+  void AllowEndPointing() {
+    KALDI_ASSERT(
+        "You must call SetOutputFrameShiftInSeconds() to use endpointing" &&
+        frame_shift_seconds_ != FLT_MAX);
```
Comment (kkm000) [inline src/cudadecoder/cuda-decoder.h:251]:
```suggestion
    if (frame_shift_seconds_ == FLT_MAX) {
      KALDI_ERR << "You must call SetOutputFrameShiftInSeconds() "
                << "to use endpointing";
    }
```

File: src/cudadecoder/cuda-decoder.h [hunk 3]
```diff
@@ -463,10 +486,16 @@ class CudaDecoder {
   CudaDecoderConfig config_;
   fst::SymbolTable *word_syms_;       // for partial hypotheses
   bool generate_partial_hypotheses_;  // set by AllowPartialHypotheses
+  bool endpointing_;
+  bool partial_traceback_;
+  BaseFloat frame_shift_seconds_;
+
+  std::set<int> silence_phones_;
```
Comment (kkm000) [inline src/cudadecoder/cuda-decoder.h:493]:
```suggestion
  std::set<int32> silence_phones_;
```

**Comment by hugovbraun** [comment]
@kkm000 The known issue was fixed and we don't have any known bugs now

**Comment by kkm000** [comment]
Thanks much! Merging.

**Comment by auzxb** [comment]
A bug happened in this commit

`cuda-decoder.h:851:3: error: ‘atomic_int32_t’ in namespace ‘std’ does not name a type
   std::atomic_int32_t n_partial_traceback_threads_todo_;`

**Comment by kkm000** [comment]
@auzxb, this is fixed in #4173 today. Please open a new issue when reporting a big, we won't notice otherwise.

**Pull request merged**